packages/muelu/src/Misc/MueLu_GeometricCoordinatesTransferFactory_def.hpp

```diff
@@ -0,0 +1,136 @@
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact
+//                    Jonathan Hu       (jhu@sandia.gov)
+//                    Andrey Prokopenko (aprokop@sandia.gov)
+//                    Ray Tuminaro      (rstumin@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+#ifndef MUELU_GEOMETRICCOORDINATESTRANSFER_FACTORY_DEF_HPP
+#define MUELU_GEOMETRICCOORDINATESTRANSFER_FACTORY_DEF_HPP
+
+#include "Xpetra_ImportFactory.hpp"
+#include "Xpetra_MultiVectorFactory.hpp"
+#include "Xpetra_VectorFactory.hpp"
+#include "Xpetra_MapFactory.hpp"
+#include "Xpetra_IO.hpp"
+#include "Xpetra_MapFactory.hpp"
+
+#include "MueLu_CoarseMapFactory.hpp"
+#include "MueLu_Aggregates.hpp"
+#include "MueLu_GeometricCoordinatesTransferFactory_decl.hpp"
+//#include "MueLu_Utilities.hpp"
+
+#include "MueLu_Level.hpp"
+#include "MueLu_Monitor.hpp"
+
+namespace MueLu {
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  RCP<const ParameterList> GeometricCoordinatesTransferFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GetValidParameterList() const {
+    RCP<ParameterList> validParamList = rcp(new ParameterList());
+
+    //validParamList->set< RCP<const FactoryBase> >("coarsening_rate", Teuchos::null, "coarsening rate for coordinates");
+    validParamList->set< std::string>            ("Coarsen",                  "2", "Coarsening rate for all spatial dimensions");
+    validParamList->set< RCP<const FactoryBase> >("Coordinates",    Teuchos::null, "Factory for coordinates generation");
+    validParamList->set< int >                   ("write start",    -1, "first level at which coordinates should be written to file");
+    validParamList->set< int >                   ("write end",      -1, "last level at which coordinates should be written to file");
+
+    return validParamList;
+  }
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void GeometricCoordinatesTransferFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::DeclareInput(Level& fineLevel, Level& coarseLevel) const {
+    static bool isAvailableCoords = false;
+    // if(fineLevel.GetLevelID()!=0) {
+    //   if (fineLevel.IsAvailable("coarsening_rate", NoFactory::get())) {
+    // 	fineLevel.DeclareInput("coarsening_rate", NoFactory::get(), this);
+    //   }
+    // }
+    // Input(fineLevel, "coarsening_rate");
+    
```

Comment:
Well, we can also store the "Coarse Coordinates" on the coarse level. We just have to make sure that the TransferFactory then "requests" data on the coarse level. maybe that's the better thing to do.